src/System.Net.Security/src/System/Net/NTAuthentication.cs

```diff
@@ -0,0 +1,677 @@
+            }
+        }
+
+        internal string ProtocolName
+        {
+            get
+            {
+                // Note: May return string.Empty if the auth is not done yet or failed.
+                if (_protocolName == null)
+                {
+                    NegotiationInfoClass negotiationInfo = null;
+
+                    if (IsValidContext)
+                    {
+                        negotiationInfo = SSPIWrapper.QueryContextAttributes(GlobalSSPI.SSPIAuth, _securityContext, Interop.SspiCli.ContextAttribute.NegotiationInfo) as NegotiationInfoClass;
+                        if (IsCompleted)
+                        {
+                            if (negotiationInfo != null)
+                            {
+                                // Cache it only when it's completed.
+                                _protocolName = negotiationInfo.AuthenticationPackage;
+                            }
+                        }
+                    }
+
+                    return negotiationInfo == null ? string.Empty : negotiationInfo.AuthenticationPackage;
+                }
+                return _protocolName;
+            }
+        }
+
+        internal SecSizes Sizes
+        {
+            get
+            {
+                GlobalLog.Assert(IsCompleted && IsValidContext, "NTAuthentication#{0}::MaxDataSize|The context is not completed or invalid.", LoggingHash.HashString(this));
+                if (_sizes == null)
+                {
+                    _sizes = SSPIWrapper.QueryContextAttributes(
+                                  GlobalSSPI.SSPIAuth,
+                                  _securityContext,
+                                  Interop.SspiCli.ContextAttribute.Sizes
+                                  ) as SecSizes;
+                }
+
+                return _sizes;
+            }
+        }
+
+        //
+        // This overload does not attempt to impersonate because the caller either did it already or the original thread context is still preserved.
+        //
+        internal NTAuthentication(bool isServer, string package, NetworkCredential credential, string spn, Interop.SspiCli.ContextFlags requestedContextFlags, ChannelBinding channelBinding)
+        {
+            Initialize(isServer, package, credential, spn, requestedContextFlags, channelBinding);
+        }
+
+        private class InitializeCallbackContext
+        {
+            internal InitializeCallbackContext(NTAuthentication thisPtr, bool isServer, string package, NetworkCredential credential, string spn, Interop.SspiCli.ContextFlags requestedContextFlags, ChannelBinding channelBinding)
+            {
+                this.thisPtr = thisPtr;
+                this.isServer = isServer;
+                this.package = package;
+                this.credential = credential;
+                this.spn = spn;
+                this.requestedContextFlags = requestedContextFlags;
+                this.channelBinding = channelBinding;
```

Comment:
`this.` should be removed.  And the internal fields should have leading underscores in their names.